api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -96,31 +97,37 @@ public Workspace get(String ns, String firecloudName) {
     List<Workspace> dbWorkspaces = workspaceDao.findAllByFirecloudUuidIn(fcWorkspaces.keySet());
 
     return dbWorkspaces.stream()
+        .filter(dbWorkspace -> dbWorkspace.getWorkspaceActiveStatusEnum() == WorkspaceActiveStatus.ACTIVE)
         .map(dbWorkspace -> {
           String fcWorkspaceAccessLevel = fcWorkspaces.get(dbWorkspace.getFirecloudUuid()).getAccessLevel();
           WorkspaceResponse currentWorkspace = new WorkspaceResponse();
           currentWorkspace.setWorkspace(workspaceMapper.toApiWorkspace(dbWorkspace));
           currentWorkspace.setAccessLevel(workspaceMapper.toApiWorkspaceAccessLevel(fcWorkspaceAccessLevel));
           return currentWorkspace;
-        }).collect(Collectors.toList());
+        })
+        .collect(Collectors.toList());
   }
 
   private Map<String, org.pmiops.workbench.firecloud.model.WorkspaceResponse> getFirecloudWorkspaces() {
-    return fireCloudService.getWorkspaces().stream().collect(
-            Collectors.toMap(
+    return fireCloudService.getWorkspaces().stream()
+            .collect(Collectors.toMap(
                 fcWorkspace -> fcWorkspace.getWorkspace().getWorkspaceId(),
                 fcWorkspace -> fcWorkspace));
   }
 
   @Override
   public Workspace getByName(String ns, String name) {
-    return workspaceDao.findByWorkspaceNamespaceAndName(ns, name);
+    Workspace workspace = workspaceDao.findByWorkspaceNamespaceAndName(ns, name);
+    if (workspace == null || workspace.getWorkspaceActiveStatusEnum() != WorkspaceActiveStatus.ACTIVE) {
+      return null;
```

Comment:
@calbach I added status checks for some of our other getter methods, however for this one (which is called to check if a ws name already exists when creating/cloning a ws) i think we do not want to error out with 404 if a workspace is found but not active, since we probably do want to allow people to create/clone workspaces with the same names as deleted workspaces, yes?

Reply:
Well.. I'm pretty confused. Our controllers seem to be acting under the assumption that workspace name is unique, yet there is no uniqueness constraint applied at the database level. In several of these cases it seems like we're expecting to return null rather than 404, which is not well documented. I think in the getByName case it should be returning null if there are no active matching workspaces.

Would need to think more about whether we want an index uniqueness constraint on name (if so, this approach of reusing the table presents some issues). If not, we should think carefully about create/clone behaviors. But I think returning null should do the trick here for now.

Reply:
Quick comment from the peanut gallery: we should definitely avoid trying to maintain any uniqueness of names unless we're otherwise forced to (e.g. notebook filenames which need to be unique within a directory).

I was going to say we should aim to use workspace IDs wherever possible in our URLs and API calls...

but it seems like Terra's workspace APIs only ever take the workspace name, not workspace ID, as parameters. In fact, it's not possible to change a workspace's name via Terra, so they must be treating it as some sort of actual ID (even though there's a separate workspaceUuid). Dang.

Still, we should probably strive to do things as "right" as possible in our own environment, and consider urging Terra to improve their APIs over time.

Reply:
This is another point of terminology confusion, I'm 99% sure firecloudName is locally unique within a project (the docs don't really say). workspaceId is a UUID which is globally unique. The name referenced here is the display name for the AoU workspace.

I don't have any context on why there's this secondary UUID "workspaceId", but I would consider the RESTFUL identifier for a Firecloud workspace to be `workspaces/{namespace}/{name}`. Our referencing that primary API identifier seems correct (assuming there aren't some strange caveats about them changing under you).

Reply:
Agreed, what we're doing re: referring to the FireCloud workspaceName as a unique & immutable identifier is correct. 

But there's still plenty of room for confusion in our codebase and on the user's end. For example, today if I create a workspace and then rename it in AoU, our workspace URL shows the old workspace name (but lowercased and with spaces removed):

<img width="477" alt="Screen Shot 2019-05-10 at 4 28 29 PM" src="https://user-images.githubusercontent.com/51842/57554978-e6469f80-7340-11e9-8612-457a22c6832e.png">

The billing model change further complicates things. We'll soon have ~5 fields all referring to some variant of the same thing:

1. GCP project ID
2. GCP project name (Terra doesn't read/write this right now, but it does show up in a user's cloud console)
3. AoU workspace name (actually a name)
4. Terra workspace name (actually an ID)
5. Terra workspace ID (a UUID)

But the change is also an opportunity – I'm optimistic that when the dust settles we can ensure / enforce that 1, 2, 4 are all identical and our URLs end up looking something like http://workbench.researchallofus.org/workspaces/aou-d2fbe9c/. Which is probably the model that Terra should work towards as well.

Reply:
:+1: I agree this would be the ideal behavior. There are definitely some issues to confront in getting there, e.g. global namespacing (a workspace name cannot match any existing Google project ID) but they are probably surmountable.